source/shared/core_conn.cpp

```diff
@@ -776,7 +778,68 @@ void determine_server_version( _Inout_ sqlsrv_conn* conn TSRMLS_DC )
     conn->server_version = version_major;
 }
 
-void common_conn_str_append_func( _In_z_ const char* odbc_name, _In_reads_(val_len) const char* val, _Inout_ size_t val_len, _Inout_ std::string& conn_str TSRMLS_DC )
+// Column Encryption feature: if a custom keystore provider is specified, 
+// load and configure it when column encryption is enabled, but this step have
+// to be executed after the connection has been established
+void load_configure_ksp( _Inout_ sqlsrv_conn* conn TSRMLS_DC )
+{
+    // If column encryption is not enabled simply do nothing. Otherwise, check if a custom keystore provider
+    // is required for encryption or decryption. Note, in order to load and configure a custom keystore provider, 
+    // all KSP fields in conn->ce_option must be defined. 
+    if ( ! conn->ce_option.enabled || ! conn->ce_option.ksp_required )
+        return;
+    
+    // Do something like the following sample
+    // use the KSP related fields in conn->ce_option
+    // CEKEYSTOREDATA is defined in msodbcsql.h
+    // https://docs.microsoft.com/en-us/sql/connect/odbc/custom-keystore-providers
+
+    CHECK_CUSTOM_ERROR( conn->ce_option.ksp_name == NULL, conn, SQLSRV_ERROR_KEYSTORE_NAME_MISSING) {
+        throw core::CoreException();
+    }
+
+    CHECK_CUSTOM_ERROR( conn->ce_option.ksp_path == NULL, conn, SQLSRV_ERROR_KEYSTORE_PATH_MISSING) {
+        throw core::CoreException();
+    }
+
+    CHECK_CUSTOM_ERROR( conn->ce_option.key_size == 0, conn, SQLSRV_ERROR_KEYSTORE_KEY_MISSING) {
+        throw core::CoreException();
+    }
+
+    char* ksp_name = Z_STRVAL_P( conn->ce_option.ksp_name );
+    char* ksp_path = Z_STRVAL_P( conn->ce_option.ksp_path );
+    unsigned int name_len = Z_STRLEN_P( conn->ce_option.ksp_name );
+    unsigned int path_len = Z_STRLEN_P( conn->ce_option.ksp_path );
+    unsigned int key_size = conn->ce_option.key_size;    
+
+    sqlsrv_malloc_auto_ptr<unsigned char> ksp_data;
+
+    ksp_data = reinterpret_cast<unsigned char*>( sqlsrv_malloc( sizeof( CEKEYSTOREDATA ) + key_size ) );
+
+    CEKEYSTOREDATA *pKsd = (CEKEYSTOREDATA*) ksp_data.get();
+
+    pKsd->dataSize = key_size;
+
+    // First, convert conn->ce_option.ksp_name to a WCHAR version 
+    unsigned int wname_len = 0;
+    sqlsrv_malloc_auto_ptr<SQLWCHAR> wksp_name;
+    wksp_name = utf16_string_from_mbcs_string( SQLSRV_ENCODING_UTF8, ksp_name, name_len, &wname_len );
+
+    CHECK_CUSTOM_ERROR( wksp_name == 0, conn, SQLSRV_ERROR_CONNECT_STRING_ENCODING_TRANSLATE ) {
+	    throw core::CoreException();
+    }
+
+    pKsd->name = (wchar_t *) wksp_name.get();
```

Comment:
Is it necessary to cast to wchar_t? wchar_t is 16 bits on Windows and 32 on Linux, and wksp_name is defined with SQLWCHAR, which should resolve to wchar_t iirc. But if not, can this lead to cross-platform incompatibility?

Reply:
Please check the keystore definitions in msodbcsql.h, and they used wchar_t in quite a few places. In Linux, when compiling the KSP source and the app, we use `-fshort-wchar`